axum-core/src/extract/mod.rs

```diff
@@ -211,27 +206,118 @@ impl<B> RequestParts<B> {
         &self.headers
     }
 
-    /// Gets a mutable reference to the request headers.
-    pub fn headers_mut(&mut self) -> &mut HeaderMap {
-        &mut self.headers
-    }
-
     /// Gets a reference to the request extensions.
     pub fn extensions(&self) -> &Extensions {
         &self.extensions
     }
 
-    /// Gets a mutable reference to the request extensions.
-    pub fn extensions_mut(&mut self) -> &mut Extensions {
-        &mut self.extensions
+    pub fn insert_extension<T>(&mut self, extension: T)
+    where
+        T: Send + Sync + 'static,
+    {
+        self.extensions.insert(extension);
     }
 
+    // TODO(david): probably want insert_header and append_header
```

Comment:
I actually think having a `remove_extension` method is also fine, since you cannot remove all extensions with such a method. Then we could make `WebSocketUpgade` implement `FromRequest<Ref, _>`.

We probably don't need `remove_header` since they you'll be able to remove all the headers, and cloning headers is always possible and cheap. You can't always clone extensions.